src/inpod/workloadmanager.rs

```diff
@@ -0,0 +1,606 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::readiness;
+use drain::Watch;
+use std::path::PathBuf;
+use std::time::Duration;
+use tokio::net::UnixStream;
+use tracing::{debug, error, info, warn};
+
+use super::statemanager::WorkloadProxyManagerState;
+use super::Error;
+
+#[mockall_double::double]
+use super::protocol::WorkloadStreamProcessor;
+
+const RETRY_DURATION: Duration = Duration::from_secs(5);
+
+struct WorkloadProxyNetworkHandler {
+    uds: PathBuf,
+}
+
+struct WorkloadProxyReadinessHandler {
+    ready: readiness::Ready,
+    // Manually drop as we don't want to mark ready if we are dropped.
+    // This can happen when the server drains.
+    block_ready: Option<std::mem::ManuallyDrop<readiness::BlockReady>>,
+}
+
+pub struct WorkloadProxyManager {
+    state: super::statemanager::WorkloadProxyManagerState,
+    networking: WorkloadProxyNetworkHandler,
+    // readiness - we are only ready when we are connected. if we get disconnected, we become not ready.
+    readiness: WorkloadProxyReadinessHandler,
+}
+
+struct WorkloadProxyManagerProcessor<'a> {
+    state: &'a mut super::statemanager::WorkloadProxyManagerState,
+    readiness: &'a mut WorkloadProxyReadinessHandler,
+
+    next_pending_retry: Option<std::pin::Pin<Box<tokio::time::Sleep>>>,
+}
+
+impl WorkloadProxyReadinessHandler {
+    fn new(ready: readiness::Ready) -> Self {
+        let mut r = Self {
+            ready,
+            block_ready: None,
+        };
+        r.not_ready();
```

Comment:
what do you mean by upgrade? can you provide the steps to reproduce this?

Reply:
just add some label in the daemonset template or similar, it will trigger a rolling restart. Real world would probably be a more useful change, of course.

Reply:
I just tried this scenario - I think you have 2 ztunnels running on the same time on the same node.

This is not supported as they both will try to open listening sockets in the same pod netns - and you'll get the error mentioned.

I believe running 2 ztunnels at the same time wasn't supported before AFAIK, so I don't believe this is a regression.

The second ztunnel should re-try binding, so as soon as the other ztunnel is removed it should come up.

We should probably change the daemonset to 
```
      maxSurge: 0
      maxUnavailable: 1
```
(today its the other way around)

Reply:
it was supported before and even fairly graceful cut over. afaik new connections went to the new one while the old one drained old connections. Without this i don't see how an upgrade is done

Reply:
I'm not sure how a graceful cutoff is possible with the old system. Do you mean old connection continued with the old ztunnel?

I'll try adding SO_REUSEPORT  to the sockets.

Reply:
Re upgrade, supporting SxS upgrades will involve more than solving the ports. So probably worth a separate discussion. I was under the impression that our initial upgrade strategy was to cordon/drain node before an upgrade

Reply:
> it was supported before and even fairly graceful cut over.

The existing support was "fairly graceful" in the sense that on a new ztunnel, it tore down all the node-level rules, recreated them pointing at the new ztunnel, and  tried to reconcile/catch any pods that might have spawned in the interim.

Doing this correctly will require more work, and yes, the only way to do this safely in production with current code in `master` is node cordon.

Reply:
> the main limitation [for SO_REUSEPORT] is that new ztunnel and old ztunnel need to have the same UID (enforced by linux for port-reuse). 

Which is not really an assumption we can make, though it may be OK for now. I'm not sure openshift will like this.

Regardless though, as before, the only current way we can recommend people to safely upgrade ztunnel in production is node cordon, so it doesn't matter.